### ISSUE TITLE
Old haguichi not detecting Terminal/FileManager etc. variables correctly

### ISSUE THREAD

**Issue by Fincer** [ztefn/haguichi#2]
Hi there,

the version of Haguichi I use is an old one, 1.0.26 clr 4.0. Why? Because I prefer GTK2 over GTK3.

I have compiled the program multiple times on my KDE5 environment in previous years without problems. I tested the program yesterday again after long time.

I noticed the program didn't correctly set neither variable _Terminal_ (DetermineTerminalThread) nor _FileManager_ (DetermineFileManagerThread) which both have been defined in _/src/Command.cs_. _Terminal_ variable always returns "gnome-terminal" and _FileManager_ returns "nautilus", though I don't have none of these installed. Instead, I use konsole and dolphin.

As far as I am aware, old haguichi config is stored at _$HOME/.config/gconf_ folder. I tried deleting it, but the problem persisted. I have enabled debug flag for the program but it didn't give any helpful output either.

Before suggesting that I should update my Haguichi version, I'd like to update the source code so that the program gets those variable correctly set or at least I can see konsole & dolphin in CommandsEditor // Settings window as expected.

You have used CSharp Exists() method for detecting a Linux command in _/src/Commands.cs_. Is this method still valid or is there something else I am not aware of? My mono version is 5.0.0.100.

If this is a source code problem, which method should be used to detect a command? If it's not a source code problem, is there any other configuration files other than those in _$HOME/.config/gconf_ folder?

Sincerely,
Fincer

**Comment by ztefn** [comment]
Hi Fincer,

You can find the Exists() method within Command.cs itself on line 235:
http://bazaar.launchpad.net/~ztefn/haguichi/1.0/view/head:/src/Command.cs#L235

As you can see it simply executes "command -v {program here}" in a bash shell. If this "command -v" doesn't find a path to the programs binary Haguichi assumes it doesn't exist.

The current Vala/GTK3 version is still using the same technique:
http://bazaar.launchpad.net/~ztefn/haguichi/1.3/view/head:/src/command.vala#L93

PS: You should update your Haguichi version! ;P

**Comment by Fincer** [comment]
Hey,

thanks for quick response @ztefn ! I'll check those lines soon.

P.S. I know. I'm just too stubborn to do that. :)

- Fincer

**Comment by ztefn** [comment]
You know you can just shrink the window down and it'll look almost the same as the GTK2 version, right?

![Screenshot](https://raw.githubusercontent.com/ztefn/gse-haguichi-indicator/master/screenshot.png)

**Comment by Fincer** [comment]
I didn't know that. Thanks for info, though. 

It's actually more that I prefer keeping my system without GTK3 packages. Qt5 is my choice, but since the program is not Qt-based, and I really like the idea of Haguichi, I use the old GTK2 version instead. 

There is a QT-based hamachi UI program known as Quamachi around, but it's just too ugly. Haguichi's interface is much better.

**Comment by ztefn** [comment]
Ah, I see... here I thought people rather kept their system without Mono packages... =)

Anyway. Please let me know if you find anything problematic with the command detection which could be relevant to the Vala code too.

**Comment by Fincer** [comment]
Yeah sure! Thanks again for responding!

**Comment by Fincer** [comment]
I manually tested the following commands (~line 235):

`bash -c "command -v konsole"`
which returns
`/usr/bin/konsole`

However, these ones return nothing on my system:
```
bash -c "command -v gnome-terminal"
bash -c "command -v nemo"_
```

which is the problem. Simply using uppercase `-V` parameter for `command` in line 238 fixes the problem.

Condition `if ( output.Contains ( "command not found" ) )` in line 240 returns _true_ because output with lowercase `-v` doesn't contain "command not found" string. It simply returns empty or null which therefore passes the check and causes strange bugs here.

**Comment by ztefn** [comment]
I'm not so sure about this. Haguichi itself makes bash return "command not found" by appending `&>/dev/null || echo 'command not found'` to the command. This prints "command not found" if the "command -v" command returns any error _**code**_ (because as you saw, it doesn't print any error _**output**_).

So these commands is what you should actually run on your system to get the same output as Haguichi:
`command -v konsole &>/dev/null || echo 'command not found'`
`command -v gnome-terminal &>/dev/null || echo 'command not found'`

And to make sure what Haguichi actually determines, please check the debug output:
`haguichi -d | grep Determine`

Because it could also be that Haguichi _does_ determine the right commands, but the threads finish only after the CommandsEditor already has been initialized. I recently stripped out the separate detection threads in the Vala code for that same reason: https://github.com/ztefn/haguichi/commit/c6e51bb82d81b5a7ba3a4e0969cf73ce24a5c34b

**Comment by Fincer** [comment]
Hmm okay. Actually, yeah...

Commands

```
command -v konsole &>/dev/null || echo 'command not found'
command -v gnome-terminal &>/dev/null || echo 'command not found'
```

work as expected.

This is the terminal output of `haguichi -d | grep Determine` without any patches to Commands.cs:

```
[02/10/2017 19:22:31 - fincer: ~ ]$ haguichi -d | grep Determine                                     
[19:23:25.4859760 ENVIRONMENT] [Command.DetermineSudoThread] [19:23:25.4859760 ENVIRONMENT] [Command.DetermineRemoteDesktopThread] Command for remote desktop: vinagre
[19:23:25.4859760 ENVIRONMENT] [Command.DetermineFileManagerThread] Command for file manager: nautilus
[19:23:25.4859760 ENVIRONMENT] [Command.DetermineTerminalThread] Command for terminal: gnome-terminal
```

I need to investigate this bit further, thanks for the hints!

**Comment by Fincer** [comment]
Removing private threads doesn't solve the issue. I added additional debug log strings for `public static bool Exists ( string command )` to see what it returns for each command, as described below.

This is my terminal output:

```
[02/10/2017 22:30:52 - fincer: ~ ]$ haguichi -d | grep Command
[22:30:54.1458050 ENVIRONMENT] [Command] pkexec found
[22:30:54.1458500 ENVIRONMENT] [Command.DetermineSudo] Command for sudo: pkexec
[22:30:54.1511830 ENVIRONMENT] [Command] gnome-terminal found
[22:30:54.1512310 ENVIRONMENT] [Command.DetermineTerminal] Command for terminal: gnome-terminal
[22:30:54.1565370 ENVIRONMENT] [Command] nautilus found
[22:30:54.1565930 ENVIRONMENT] [Command.DetermineFileManager] Command for file manager: nautilus
[22:30:54.1621170 ENVIRONMENT] [Command] vinagre not found
[22:30:54.1659930 ENVIRONMENT] [Command] krdc not found
[22:30:54.1660480 ENVIRONMENT] [Command.DetermineRemoteDesktop] Command for remote desktop: 
[22:30:54.1714710 ENVIRONMENT] [Command] hamachi found
[22:30:54.1801770 ENVIRONMENT] [Command] hamachi found
[22:30:54.1923570 ENVIRONMENT] [Command] systemctl found
```

_gnome-terminal_ and _nautilus_ should return "not found". Others, such as _vinagre_ and _krdc_ are OK.

And this is the code I added, lines 210-221:

```
        if (output.Contains ( "command not found" ) )
        {
            Debug.Log ( Debug.Domain.Environment, "Command", command + " not found" );

            return false;
        }
        else
        {
            Debug.Log ( Debug.Domain.Environment, "Command", command + " found" );
            
            return true;
        }
```

**Comment by ztefn** [comment]
Might be interesting to also log the output that Haguichi receives, right before the if/else statement:
`Debug.Log ( Debug.Domain.Environment, "Command", "Output for command " + command + ": " + output );`

**Comment by Fincer** [comment]
Yup. I added `Debug.Log`s here and there. 

`Exists` never returned a correct boolean value (_false_ in my case) for gnome-terminal command. For nautilus and dolphin...it returned _true_ sometimes, sometimes _false_.

`ReturnOutput` (lines 297-338) never returned "command not found" for _gnome-terminal_. Instead, it returned a null/empty value for it. 

All randomness and strange outputs stopped once I added _p.WaitForExit ();_. I didn't even need to delete private threads you mentioned before.

Sorry for having troubled you with this issue.
